src/mds/Server.cc

```diff
@@ -4106,6 +4107,17 @@ void Server::handle_client_getattr(const MDRequestRef& mdr, bool is_lookup)
     if (r > 0)
       return; // delayed
 
+    // Do not batch if any xlock is held
+    if (!r) {
+      CInode *in = mdr->in[0];
+      if (((mask & CEPH_CAP_LINK_SHARED) && (in->linklock.is_xlocked_by_client(client))) ||
+          ((mask & CEPH_CAP_AUTH_SHARED) && (in->authlock.is_xlocked_by_client(client))) ||
+          ((mask & CEPH_CAP_XATTR_SHARED) && (in->xattrlock.is_xlocked_by_client(client))) ||
+          ((mask & CEPH_CAP_FILE_SHARED) && (in->filelock.is_xlocked_by_client(client)))) {
+	r = -1;
```

Comment:
There is another bug I think. This won't handle the case where we're skipping e.g. a rdlock when the client is issued an exclusive cap, like `Ax`. Consider:

client 1: issued `pAx`
client 2: `getattr pAsLsXsFs` issued `pLxXx`
client 3: `getattr pAsLsXsFs` issued `p`

client 2 will skip acquiring the linklock/xattrlock because it's issued `LxXx`. It will block on recall of `Ax` from client 1. 

Client 2's `getattr` cannot be a batch head for client 3's `getattr.

This all exposes a problem I believe exists with where we're constructing the batch head: this should be constructed only after this `acquire_locks` fails:

https://github.com/ceph/ceph/blob/2c16096b569ed13eaf145491671bbe9eec90ee92/src/mds/Server.cc#L4217-L4218

Before that point, record whether we've skipped any locks due to issued caps or (for this particular bug) the lock is already xlocked by the client. The latter cannot be easily checked without looking at the inode's lock because the Locker state machine hides why a rdlock succeeds (in this case, the client has an xlock already).

Reply:
> There is another bug I think. This won't handle the case where we're skipping e.g. a rdlock when the client is issued an exclusive cap, like `Ax`. Consider:
> 
> client 1: issued `pAx` client 2: `getattr pAsLsXsFs` issued `pLxXx` client 3: `getattr pAsLsXsFs` issued `p`
> 
@batrick Correct me if I am wrong here.

From `mds/lock.c` I can see that only the loner client could get the `x` caps. Could you point me out in which case could we see the `x` in different locks will be issued to different clients at the same time ?

Reply:
> > There is another bug I think. This won't handle the case where we're skipping e.g. a rdlock when the client is issued an exclusive cap, like `Ax`. Consider:
> > client 1: issued `pAx` client 2: `getattr pAsLsXsFs` issued `pLxXx` client 3: `getattr pAsLsXsFs` issued `p`
> 
> @batrick Correct me if I am wrong here.
> 
> From `mds/lock.c` I can see that only the loner client could get the `x` caps. Could you point me out in which case could we see the `x` in different locks will be issued to different clients at the same time ?

You might be right; I'm not sure. In principle I don't see why it wouldn't be allowed but the state diagram suggests it is not. 

> Different locks could have different loners at the same time ?

It doesn't look like it but it's worth checking.

I think the batch leader construction should still be moved however. And that "detail" of the loner client shouldn't be relied on for the checks in any case.

Reply:
> > > There is another bug I think. This won't handle the case where we're skipping e.g. a rdlock when the client is issued an exclusive cap, like `Ax`. Consider:
> > > client 1: issued `pAx` client 2: `getattr pAsLsXsFs` issued `pLxXx` client 3: `getattr pAsLsXsFs` issued `p`
> > 
> > 
> > @batrick Correct me if I am wrong here.
> > From `mds/lock.c` I can see that only the loner client could get the `x` caps. Could you point me out in which case could we see the `x` in different locks will be issued to different clients at the same time ?
> 
> You might be right; I'm not sure. In principle I don't see why it wouldn't be allowed but the state diagram suggests it is not.
> 
> > Different locks could have different loners at the same time ?
> 
> It doesn't look like it but it's worth checking.
> 
> I think the batch leader construction should still be moved however. And that "detail" of the loner client shouldn't be relied on for the checks in any case.

I couldn't remember I ever saw this case, just now I checked some debug logs, such as the debug logs from:

```ceph-post-file: fb9a96f9-5f6d-46b4-b1fa-8580928b2241```

I didn't find any case will do this. And also by going through the mds code, only when a longer is successfully set will the `EXCL` locker state could be set. And all the lockers in a CInode will be set to the same single longer.

Reply:
Then perhaps the better check is: "is this client the loner for the CInode? In that case, do not make it a batch head".

Let's move the batch code below this call to `acquire_locks` on failure:

https://github.com/ceph/ceph/blob/2c16096b569ed13eaf145491671bbe9eec90ee92/src/mds/Server.cc#L4217-L4218

If we fail to acquire the locks, then make it the batch head if one does not exist. If a batch head does exist already, then drop locks and add it to the batch queue.

Reply:
> Then perhaps the better check is: "is this client the loner for the CInode? In that case, do not make it a batch head".
> 

I am afraid this won't work. Can we be sure that when a `client` has acquired the `xlock` will it always be the `CInode's` loner ?

Such as when a client only has the `Fs` caps and when it will send a sync truncate request to MDS, and then the truncated request in MDS will try to acquire the `xlock`. There could be no any loner. 


> Let's move the batch code below this call to `acquire_locks` on failure:
> 
> https://github.com/ceph/ceph/blob/2c16096b569ed13eaf145491671bbe9eec90ee92/src/mds/Server.cc#L4217-L4218
> 
> If we fail to acquire the locks, then make it the batch head if one does not exist. If a batch head does exist already, then drop locks and add it to the batch queue.

Sure.

Reply:
> > Let's move the batch code below this call to `acquire_locks` on failure:
> > https://github.com/ceph/ceph/blob/2c16096b569ed13eaf145491671bbe9eec90ee92/src/mds/Server.cc#L4217-L4218
> > 
> > If we fail to acquire the locks, then make it the batch head if one does not exist. If a batch head does exist already, then drop locks and add it to the batch queue.
> 
> Sure.

Checked the code carefully again, if my understanding is correct this won't work well as expected ?

For example, if the first `lookup` request just wants to acquire the `rdlock` for `linklock` and it succeeds, and then later the second `lookup` request comes and also just wants to the `rdlock` for the `linklock`. If both these two requests succeeds there won't be any chance to batch them. Actually we should batch them, right ?

Reply:
There are no further waits after that `acquire_locks` so I don't think so?

Reply:
Yeah, I think you should right here. Let me check it more.

Reply:
@batrick  If we will move the batch code after `acquire_locks`, we also should adjust the `acquire_locks` and other callers to make sure that they won't add the request to any waiter, which will retry the request later.

Reply:
Hm, that's right. Let's leave it this way then.